src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -168,44 +261,55 @@ private SafeCurlHandle CreateRequestHandle(RequestCompletionSource state)
+            if (state.RequestMessage.Content != null)
+            {
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_UPLOAD, 1L);
+            }
+            if (state.RequestMessage.Method == HttpMethod.Head)
+            {
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_NOBODY, 1L);
             }
 
-            // TODO: Handle headers, proxy and other options
-
-            return requestHandle;
-        }
-
-        private HttpResponseMessage CreateResponseMessage(SafeCurlHandle requestHandle, HttpRequestMessage request)
-        {
-            var response = new HttpResponseMessage();
-            long httpStatusCode = 0;
-            int result = Interop.libcurl.curl_easy_getinfo(requestHandle, CURLINFO.CURLINFO_RESPONSE_CODE,
-                ref httpStatusCode);
-            if (CURLcode.CURLE_OK != result)
+            // Setup callbacks
+            SetCurlOption(requestHandle, CURLoption.CURLOPT_HEADERDATA, GCHandle.ToIntPtr(stateHandle));
+            SetCurlOption(requestHandle, CURLoption.CURLOPT_HEADERFUNCTION, (Interop.libcurl.curl_readwrite_callback)CurlReceiveHeadersCallback);
+            if (state.RequestMessage.Method != HttpMethod.Head)
             {
-                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_WRITEDATA, GCHandle.ToIntPtr(stateHandle));
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_WRITEFUNCTION, (Interop.libcurl.curl_readwrite_callback)CurlReceiveBodyCallback);
+            }
+            if (state.RequestMessage.Content != null)
+            {
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_READDATA, GCHandle.ToIntPtr(stateHandle));
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_READFUNCTION, (Interop.libcurl.curl_readwrite_callback)CurlSendCallback);
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_IOCTLDATA, GCHandle.ToIntPtr(stateHandle));
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_IOCTLFUNCTION, (Interop.libcurl.curl_ioctl_callback)CurlSendIoCtlCallback);
             }
-            response.StatusCode = (HttpStatusCode) httpStatusCode;
 
-            // TODO: Do error processing if needed and return actual response
-            response.Content =
-                new StringContent("SendAsync to " + request.RequestUri.OriginalString + " returned: " +
-                                  response.StatusCode);
-            return response;
+            // TODO: Handle proxy and other options
+
+            headerPtr = SetRequestHeaders(requestHandle, state);
+
+            return requestHandle;
         }
 
         private void HandleAsyncException(RequestCompletionSource state, Exception ex)
         {
-            if (ex is OperationCanceledException)
+            if (state.SavedException != null)
+            {
+                ex = state.SavedException;
```

Comment:
Yes.. priority is given to any exception that happened in the callback since it will surely trigger the "outer" exception by failing curl_easy_perform